internal/internal_worker.go

```diff
@@ -527,155 +525,218 @@ func (aw *activityWorker) Stop() {
 }
 
-func (th *hostEnvImpl) RegisterWorkflowWithOptions(
+func (r *registry) RegisterWorkflowWithOptions(
 	af interface{},
 	options RegisterWorkflowOptions,
-) error {
+) {
 	// Validate that it is a function
 	fnType := reflect.TypeOf(af)
 	if err := validateFnFormat(fnType, true); err != nil {
-		return err
+		panic(err)
 	}
 	fnName := getFunctionName(af)
 	alias := options.Name
 	registerName := fnName
 	if len(alias) > 0 {
 		registerName = alias
 	}
-	// Check if already registered
-	if _, ok := th.getWorkflowFn(registerName); ok {
-		return fmt.Errorf("workflow name \"%v\" is already registered", registerName)
+	if !options.DisableAlreadyRegisteredCheck {
+		if _, ok := r.workflowFuncMap[registerName]; ok {
+			panic(fmt.Sprintf("workflow name \"%v\" is already registered", registerName))
+		}
 	}
-	th.addWorkflowFn(registerName, af)
+	r.addWorkflowFn(registerName, af)
 	if len(alias) > 0 {
-		th.addWorkflowAlias(fnName, alias)
+		r.addWorkflowAlias(fnName, alias)
 	}
-	return nil
 }
 
-func (th *hostEnvImpl) RegisterActivity(af interface{}) error {
-	return th.RegisterActivityWithOptions(af, RegisterActivityOptions{})
+func (r *registry) RegisterActivity(af interface{}) {
+	r.RegisterActivityWithOptions(af, RegisterActivityOptions{})
 }
 
-func (th *hostEnvImpl) RegisterActivityWithOptions(
+func (r *registry) RegisterActivityWithOptions(
 	af interface{},
 	options RegisterActivityOptions,
-) error {
+) {
 	// Validate that it is a function
 	fnType := reflect.TypeOf(af)
+	if fnType.Kind() == reflect.Ptr && fnType.Elem().Kind() == reflect.Struct {
+		r.registerActivityStructWithOptions(af, options)
+		return
+	}
 	if err := validateFnFormat(fnType, false); err != nil {
-		return err
+		panic(err)
 	}
 	fnName := getFunctionName(af)
 	alias := options.Name
 	registerName := fnName
 	if len(alias) > 0 {
 		registerName = alias
 	}
-	// Check if already registered
-	if _, ok := th.getActivityFn(registerName); ok {
-		return fmt.Errorf("activity type \"%v\" is already registered", registerName)
+	if !options.DisableAlreadyRegisteredCheck {
+		if _, ok := r.activityFuncMap[registerName]; ok {
+			panic(fmt.Sprintf("activity type \"%v\" is already registered", registerName))
+		}
 	}
-	th.addActivityFn(registerName, af)
+	r.addActivityFn(registerName, af)
 	if len(alias) > 0 {
-		th.addActivityAlias(fnName, alias)
+		r.addActivityAlias(fnName, alias)
+	}
+}
+
+func (r *registry) registerActivityStructWithOptions(aStruct interface{}, options RegisterActivityOptions) error {
+	structValue := reflect.ValueOf(aStruct)
+	structType := structValue.Type()
+	count := 0
+	for i := 0; i < structValue.NumMethod(); i++ {
+		methodValue := structValue.Method(i)
+		method := structType.Method(i)
+		// skip private method
+		if method.PkgPath != "" {
+			continue
+		}
+		name := method.Name
+		if err := validateFnFormat(method.Type, false); err != nil {
+			return fmt.Errorf("method %v of %v: %e", name, structType.Name(), err)
```

Comment:
In line 582, we panic if the validation fails, do you want to do the same here and the below already register check?

Reply:
Yes good catch. Fixed.